third_party/parser/include/ruby_parser/driver.hh

```diff
@@ -3,169 +3,164 @@
+    }
+
+    template <typename... Args> ruby_parser::delimited_node_list *delimited_node_list(Args &&... args) {
+        return _delimited_node_list.alloc(std::forward<Args>(args)...);
+    }
+
+    template <typename... Args> ruby_parser::delimited_block *delimited_block(Args &&... args) {
+        return _delimited_block.alloc(std::forward<Args>(args)...);
+    }
+
+    template <typename... Args> ruby_parser::node_with_token *node_with_token(Args &&... args) {
+        return _node_with_token.alloc(std::forward<Args>(args)...);
+    }
+
+    template <typename... Args> ruby_parser::case_body *case_body(Args &&... args) {
+        return _case_body.alloc(std::forward<Args>(args)...);
+    }
 };
 
 class base_driver {
 public:
-	diagnostics_t diagnostics;
-	const builder& build;
-	lexer lex;
-	mempool alloc;
-
-	bool pending_error;
-	size_t def_level;
-	ForeignPtr ast;
-
-	base_driver(ruby_version version, const std::string& source, const struct builder& builder);
-	virtual ~base_driver() {}
-	virtual ForeignPtr parse(SelfPtr self) = 0;
-
-	bool valid_kwarg_name(const token *name) {
-		char c = name->string().at(0);
-		return !(c >= 'A' && c <= 'Z');
-	}
-
-	ruby_parser::state_stack *copy_stack() {
-		return alloc._stacks.alloc(lex.cmdarg);
-	}
-
-	void replace_stack(ruby_parser::state_stack *stack) {
-		lex.cmdarg = *stack;
-	}
-
-	void external_diagnostic(dlevel lvl, dclass cls, size_t begin, size_t end, const std::string &msg) {
-		diagnostics.emplace_back(lvl, cls, diagnostic::range(begin, end), msg);
-		if (lvl == dlevel::ERROR) {
-			pending_error = true;
-		}
-	}
+    diagnostics_t diagnostics;
+    const builder &build;
+    lexer lex;
+    mempool alloc;
+
+    bool pending_error;
+    size_t def_level;
+    ForeignPtr ast;
+    token_t last_token;
+
+    base_driver(ruby_version version, const std::string &source, const struct builder &builder);
+    virtual ~base_driver() {}
+    virtual ForeignPtr parse(SelfPtr self) = 0;
+
+    bool valid_kwarg_name(const token *name) {
+        char c = name->string().at(0);
+        return !(c >= 'A' && c <= 'Z');
+    }
+
+    ruby_parser::state_stack *copy_stack() {
+        return alloc._stacks.alloc(lex.cmdarg);
+    }
+
+    void replace_stack(ruby_parser::state_stack *stack) {
+        lex.cmdarg = *stack;
+    }
+
+    void external_diagnostic(dlevel lvl, dclass cls, size_t begin, size_t end, const std::string &msg) {
+        diagnostics.emplace_back(lvl, cls, diagnostic::range(begin, end), msg);
+        if (lvl == dlevel::ERROR) {
+            pending_error = true;
+        }
+    }
 };
 
 class typedruby25 : public base_driver {
 public:
-	typedruby25(const std::string& source, const struct builder& builder);
-	virtual ForeignPtr parse(SelfPtr self);
-	~typedruby25() {}
+    typedruby25(const std::string &source, const struct builder &builder);
+    virtual ForeignPtr parse(SelfPtr self);
+    ~typedruby25() {}
 };
 
-}
+} // namespace ruby_parser
 
 #endif
```

Comment:
It's pretty difficult to tell what changed in this file given the amount of formatting changes. Was this just re-formatting, or are there semantic changes as well?

Reply:
only 1 line has changed: `token_t last_token`, the rest is just an auto-formatting

Reply:
Thanks!